thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/TriggerConditionGrouper.java

```diff
@@ -0,0 +1,283 @@
+  private List<MergedAnomalyResultDTO> andGrouping(
+      List<MergedAnomalyResultDTO> anomalyListA, List<MergedAnomalyResultDTO> anomalyListB) {
+    Set<MergedAnomalyResultDTO> groupedAnomalies = new HashSet<>();
+    List<MergedAnomalyResultDTO> anomalies = mergeAndSortAnomalies(anomalyListA, anomalyListB);
+
+    Stack<MergedAnomalyResultDTO> anomalyStack = new Stack<>();
+    for (MergedAnomalyResultDTO anomaly : anomalies) {
+      if (anomalyStack.isEmpty()) {
+        anomalyStack.push(makeParentEntityAnomaly(anomaly));
+        continue;
+      }
+
+      MergedAnomalyResultDTO anomalyStackTop = anomalyStack.pop();
+
+      if (anomaly.getStartTime() > anomalyStackTop.getEndTime()) {
+        // No overlap
+        anomalyStack.push(makeParentEntityAnomaly(anomaly));
+      } else if (anomaly.getEndTime() > anomalyStackTop.getEndTime()) {
+        //Partial overlap
+        anomalyStackTop.setStartTime(anomaly.getStartTime());
+        setEntityChildMapping(anomalyStackTop, anomaly);
+
+        anomalyStack.push(anomalyStackTop);
+        anomalyStack.push(makeParentEntityAnomaly(anomaly));
+      } else {
+        // Complete overlap
+        MergedAnomalyResultDTO temp = makeEntityCopy(anomalyStackTop);
+
+        anomalyStackTop.setStartTime(anomaly.getStartTime());
+        anomalyStackTop.setEndTime(anomaly.getEndTime());
+        setEntityChildMapping(anomalyStackTop, anomaly);
+
+        anomalyStack.push(anomalyStackTop);
+        anomalyStack.push(temp);
+      }
+    }
+
+    // Pop the extra anomaly out
+    if (!anomalyStack.isEmpty()) {
+      anomalyStack.pop();
+    }
+
+    // Add all parent anomalies along with their children
+    groupedAnomalies.addAll(anomalyStack);
+    for (MergedAnomalyResultDTO anomaly : anomalyStack) {
+      groupedAnomalies.addAll(anomaly.getChildren());
+    }
+
+    return new ArrayList<>(groupedAnomalies);
+  }
+
+  /**
+   * Group based on 'OR' criteria - Entity has anomaly if either sub-entity A or B have anomalies.
+   * This means we find the total anomaly coverage.
+   *
+   * Since the anomalies from the respective entities/metrics are merged
+   * before calling the grouper, we do not have to deal with overlapping
+   * anomalies within an entity/metric
+   *
+   * Sort anomalies by start time and iterate through the list
+   * Create new entity anomaly with first anomaly as child and push to stack
+   * For each anomaly following, pop stack and compare
+   * 1. No overlap:
+   *    If the anomaly doesn't overlap, push popped anomaly, create new entity anomaly with current as child and push
+   * 2. Partial overlap:
+   *    If partial overlap, update popped anomaly's end time, child and push to stack
+   * 3. Full overlap:
+   *    If full overlap, update popped anomaly with child info and push it back to stack
+   * Note the each time an entity copy is created, the children needs to be marked and set appropriately
+   *
+   */
+  private List<MergedAnomalyResultDTO> orGrouping(
```

Comment:
Could be simplified as  below: 

 private List<MergedAnomalyResultDTO> orGrouping(
      List<MergedAnomalyResultDTO> anomalyListA, List<MergedAnomalyResultDTO> anomalyListB) {

     List<MergedAnomalyResultDTO> anomalies = mergeAndSortAnomalies(anomalyListA, anomalyListB);
     if (anomalyListA.isEmpty() && anomalyListB.isEmpty()) return anomalies;

     MergedAnomalyResultDTO cur = makeParentEntityAnomaly(anomalies.get(0));
     for (int i = 1; i < anomalies.size();  i++) {
       if (anomalies.get(i). getStartTime() <= cur.getEndTime())) {
            cur. setEndTime(Math.max(anomalies.get(i). getEndTime), cur.getEndTime());
            setEntityChildMapping(cur, anomalies.get(i);
       }  else {
            anomalies.add(cur);
            cur = makeParentEntityAnomaly(anomalies.get(i));
       }
     }
    anomalies.add(cur);
    return anomalies;
}

Reply:
This is indeed much cleaner. Updated accordingly.